### ISSUE TITLE
My Mercurial Setup and Workflow at Khan Academy

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4543]
My Mercurial Setup and Workflow at Khan Academy<br>
<br>
http://bit.ly/N8Ucjb<br>
<br>
Logo
Created using Figma<br>
<br>
<div><article><h1>My Mercurial Setup and Workflow at Khan Academy</h1> <p>I’ve been using mercurial full time during my internship at <a href="http://www.khanacademy.org">Khan Academy</a> now for just over two months. I spent a good chunk of my first few weeks here in on IRC in #mercurial sharpening my new tools and more or less making my mercurial setup work identically to my git setup.</p> <p>In this post, I’ll be detailing both my tooling and my amend/rebase workflow with mercurial. For starters, make sure you have the latest version of mercurial, or you’ll run into issues following this. At the time of this post, I’m using 2.2.2. You can check what version you have with <code>hg --version</code>.</p> <p>If you just want to read code, and not words, I have my dotfiles hosted on github, which you can see here: <a href="https://github.com/jlfwong/dotfiles/">jlfwong/dotfiles</a>. Fellow Khan Academy intern <a href="http://david-hu.com/">David Hu</a> created a stripped down set of dotfiles specifically for KA that will be much easier to follow, which you can see here: <a href="https://github.com/khan/khan-dotfiles/">khan/khan-dotfiles</a>.</p> <p>This post assumes you are reasonably familiar with mercurial. If not, I suggest reading <a href="http://hgbook.red-bean.com/">Mercurial: The Definitive Guide</a> by Bryan O’Sullivan. You can read it for free online.</p> <p><nav id="TableOfContents"></nav></p> <h1 id="extensions">Extensions</h1> <p>Mercurial ships with a powerful plugin system to extend its capabilities. Quite honestly, it would be no fun at all to use without some of these extensions.</p> <p>There are two kinds of extensions for mercurial - those that are built-in, and those that require you to download something. The built-in ones ship with mercurial, but are disabled by default. Enabling extensions is done by editing the <code>[extensions]</code> section of your <code>~/.hgrc</code> file. If you want to mirror the setup described in this article, <code>~/.hgrc</code> should contain something that looks like this at the end:</p> <pre><code>[extensions] color = graphlog = pager = rebase = purge = remotebranches = ~/path/to/hg_remotebranches.py shelve = ~/path/to/hgshelve.py prompt = ~/path/to/hg-prompt/prompt.py </code></pre> <p>The color, graphlog, pager and rebase extensions are built-in, so you don’t have to specify a path for them</p> <p>I actually have more extensions than this installed, but these are the ones I use on a regular basis.</p> <h2 id="color">color</h2> <p><a href="http://mercurial.selenic.com/wiki/ColorExtension">http://mercurial.selenic.com/wiki/ColorExtension</a> (builtin)</p> <p>Simply put - this enables color in the mercurial command-line.</p> <p><img src="http://jamie-wong.com/images/12-07-11-hg-color.png" alt="Color Extension"></p> <p>This is useful in a variety of commands. Shown above is <code>hg log --patch</code> with and without color enabled. The color extension also provides color support for <code>status</code>, <code>resolve</code> and <code>status</code>.</p> <h2 id="graphlog">graphlog</h2> <p><a href="http://mercurial.selenic.com/wiki/GraphlogExtension/">http://mercurial.selenic.com/wiki/GraphlogExtension/</a> (builtin)</p> <p>The default output of <code>hg log</code> isn’t very helpful, especially if you have a non-linear history. For instance - we’ll take a look at 4 commits in our repo. Here’s the output of <code>hg log</code>:</p> <p><img src="http://jamie-wong.com/images/12-07-11-hg-log.png" alt="hg log"></p> <p>It just shows you commits in reverse order of when they entered the repository.<br> It’s much more useful to be able to see the graph of all commits. The graphlog extension enables that - giving us the slightly more useful log output. Here’s the output of <code>hg glog</code>:</p> <p><img src="http://jamie-wong.com/images/12-07-11-hg-glog.png" alt="hg graphlog"></p> <p>Now we can at least see that 3 commits that were listed sequentially by <code>hg log</code> are not actually linear in the commit history. You can also see that commit 13048 is diverged from the other commits shown.</p> <p>The problem with both of these is that for the most part, we don’t care about the majority of the information here, and there’s useful information missing. Thankfully, mercurial supports custom log formats by specifying templates of sorts. (See <a href="http://mercurial.selenic.com/wiki/Theming">Mercurial Theming</a>.)</p> <p>I already had a log format I liked from working with git, so all I had to do was convert it to mercurial. Here’s the output of my custom <code>hg lga</code>:</p> <p><img src="http://jamie-wong.com/images/12-07-11-hg-lga.png" alt="hg lga"></p> <p>This contains a bunch of extra information in a much denser format, and pretty much all of that information is useful.</p> <p>Information provided left to right:</p> <ul>
<li>Short revision id (to get the full SHA1, run <code>hg lga -v</code> once you have <code>hg lga</code> set up)</li> <li>The commit phase. I use phases to tell me whether or not a commit has been pushed. See <a href="http://mercurial.selenic.com/wiki/Phases">Mercurial Phases</a>.</li> <li>Author of the commit</li> <li>First line of commit message</li> <li>How long it’s been since it was committed. When I want to know exactly when something was committed, I do <code>hg log --stat --color=always -vpr 13049</code>, which I have aliased to <code>hg show 13049</code>.</li> <li>Any bookmarks or tags on the commit (more on bookmarks and tags later).</li> </ul>
<p>To get your log to look like this, install the graphlog extension, then create a file called <code>map-cmdline.lg</code>, and put this in it:</p> <div><pre><code><span></span><span>changeset</span> <span>=</span> <span>'</span><span>\033</span><span>[0;31m{rev}:{phase} </span><span>\033</span><span>[0;34m{author|person}</span><span>\033</span><span>[0m </span> <span>{</span><span>desc</span><span>|</span><span>firstline</span><span>|</span><span>strip</span><span>}</span> \<span>033</span><span>[</span><span>0</span><span>;</span><span>32</span><span>m</span><span>({</span><span>date</span><span>|</span><span>age</span><span>})</span> <span>{</span><span>branches</span><span>}{</span><span>bookmarks</span><span>}{</span><span>tags</span><span>}</span>\<span>n</span>\<span>n</span><span>'</span> <span>changeset_verbose</span> <span>=</span> <span>'</span><span>\033</span><span>[0;31m{rev}:{node|short}:{phase} </span><span>\033</span><span>[0;34m{author|person}</span><span>\033</span><span>[0m {desc|firstline|strip} </span><span>\033</span><span>[0;32m({date|age}) {branches}{bookmarks}{tags}</span><span>\n\n</span><span>'</span> <span>start_branches</span> <span>=</span> <span>' '</span> <span>branch</span> <span>=</span> <span>'</span><span>\033</span><span>[0;31m{branch}</span><span>\033</span><span>[0m'</span> <span>start_bookmarks</span> <span>=</span> <span>' '</span> <span>bookmark</span> <span>=</span> <span>'</span><span>\033</span><span>[0;31m[{bookmark}]</span><span>\033</span><span>[0m '</span> <span>last_bookmark</span> <span>=</span> <span>'</span><span>\033</span><span>[0;31m[{bookmark}]</span><span>\033</span><span>[0m'</span> <span>start_tags</span> <span>=</span> <span>' '</span> <span>tag</span> <span>=</span> <span>'</span><span>\033</span><span>[0;31m{tag}</span><span>\033</span><span>[0m, '</span> <span>last_tag</span> <span>=</span> <span>'</span><span>\033</span><span>[0;31m{tag}</span><span>\033</span><span>[0m'</span> </code></pre></div> <p>then add this alias to your <code>~/.hgrc</code>, correcting the path to <code>map-cmdline.lg</code>.</p> <div><pre><code><span></span><span>[alias]</span> <span>lga</span> <span>=</span> <span>glog --style=your/path/to/map-cmdline.lg</span> </code></pre></div> <p>If you want a similar log for git, add this to your <code>~/.gitconfig</code>.</p> <div><pre><code><span></span><span>[alias]</span> <span>lg</span> <span>=</span> <span>log --graph --pretty=format:'%Cred%h%Creset %Creset%Cblue%an%Creset </span> <span>%s %Cgreen(%cr)%Cred%d%Creset' --abbrev-commit --date</span><span>=</span><span>relative</span> </code></pre></div> <p><a href="http://mercurial.selenic.com/wiki/PagerExtension">http://mercurial.selenic.com/wiki/PagerExtension</a> (builtin)</p> <p>By default, when you run mercurial commands that have a big output, they just vomit everything onto the screen. This is especially annoying for <code>hg log</code>, since it will dump the entire repository history onto your screen.</p> <p>The pager extension fixes this by piping the output of selected commands through a pager of your choice. After enabling the plugin in your <code>.hgrc</code>, add the following pager configuration section:</p> <div><pre><code><span></span><span>[pager]</span> <span>pager</span> <span>=</span> <span>LESS='FSRX' less</span> <span>attend</span> <span>=</span> <span>cat, diff, glog, log, incoming, outgoing, lg, show, lga</span> </code></pre></div> <p>This specifies which command to use as a pager, and which commands hg commands you want to be piped through the pager.</p> <h2 id="rebase">rebase</h2> <p><a href="http://mercurial.selenic.com/wiki/RebaseExtension">http://mercurial.selenic.com/wiki/RebaseExtension</a> (builtin)</p> <p>The rebase extension provides a direct equivalent of <code>git rebase</code> for mercurial. For more information on why I rebase instead of merge, see below and the see my past post <a href="http://jamie-wong.com/2012/05/25/an-argument-for-mutable-local-history/">An Argument for Mutable Local History</a>.</p> <p>There isn’t really any configuration required for this one.</p> <h2 id="purge">purge</h2> <p><a href="http://mercurial.selenic.com/wiki/PurgeExtension">http://mercurial.selenic.com/wiki/PurgeExtension</a> (builtin)</p> <p>The purge extension lets you clean your working directory of all untracked files. This is often useful when reviewing changes that create new files.</p> <p>To completely reset the state of your working copy to what it was at the latest commit, I have an alias that removes untracked files and reverts all changes in the working copy:</p> <div><pre><code><span></span><span>[alias]</span> <span># Nuke everything in the working copy</span> <span>nuke</span> <span>=</span> <span>!hg up -C . &amp;&amp; hg clean</span> </code></pre></div> <h2 id="remotebranches">remotebranches</h2> <p><a href="http://mercurial.selenic.com/wiki/RemoteBranchesExtension">http://mercurial.selenic.com/wiki/RemoteBranchesExtension</a> (requires download)</p> <p>It’s often useful to have a quick answer to the question “what commit is the remote on?“. You can infer this information using <code>hg outgoing</code>, but it’s much nicer to be able to see this information marked in the revision history. This is also particularly useful because I frequently need to switch to this commit either to review others’ work or to start a new feature.</p> <p>The remotebranches extension provides this by attaching tags to commits whenever you push or pull. So after you pull from the default remote, you’ll have a commit tagged with <code>default/default</code>. Then switching to this commit becomes as easy as running <code>hg up default/default</code>.</p> <p>Unlike the previous extensions, this one requires a download. After you’ve downloaded the plugin, you need to specify the path to where you downloaded it. So you’ll have a line like this in your <code>~/.hgrc</code>, under the <code>[extensions]</code> header:</p> <div><pre><code><span></span><span>remotebranches</span> <span>=</span> <span>~/path/to/hg_remotebranches.py</span> </code></pre></div> <h2 id="shelve">shelve</h2> <p><a href="http://mercurial.selenic.com/wiki/ShelveExtension">http://mercurial.selenic.com/wiki/ShelveExtension</a> (requires download)</p> <p>This is a rough equivalent of <code>git stash</code>. For those unfamiliar, it allows you to temporarily store the changes to your working copy without making a commit.<br> This means you can clean your working directory to test something quickly, the bring the changes back easily.</p> <p><code>hg shelve</code> stores the changes in your working copy, <code>hg unshelve</code> brings those changes back.</p> <p>The behaviour of this is significantly different than <code>git stash</code> in that it prompts you for file-by-file changes to save, provides the ability to name multiple shelves with <code>hg shelve NAME</code>, and does not operate on a stack like <code>git stash</code> does.</p> <h2 id="prompt">prompt</h2> <p><a href="http://mercurial.selenic.com/wiki/PromptExtension">http://mercurial.selenic.com/wiki/PromptExtension</a> (requires download)</p> <p>Having contextual information about the repository in your prompt is really useful for seeing the status at a glance. My bash prompt when I’m working in a mercurial repository looks like this:</p> <p><img src="http://jamie-wong.com/images/12-07-11-hg-prompt.png" alt="hg prompt"></p> <p>After getting the plugin and installing it, you need to inject it into your shell prompt. You can do that by adding something like this to your <code>~/.bash_profile</code>:</p> <div><pre><code><span></span><span>function</span> hg_ps1 <span>{</span> hg prompt <span>"({root|basename}@{bookmark}{[{tags|quiet|,}]}{ {status|modified|unknown}}{ {shelf}}) "</span> <span>2</span>&gt; /dev/null <span>}</span> <span>export</span> <span>PS1</span><span>=</span><span>'\e[33m\][\W] \e[31m\]$(hg_ps1)\e[0m\]'</span> </code></pre></div> <p>This displays, from left to right:</p> <ul>
<li>the directory name of the repository root</li> <li>the current bookmark I’m on</li> <li>any tags other than <code>tip</code>
</li> <li>the status of the working copy - whether or not there are any uncommitted changes or untracked files</li> <li>the status of any shelves (this one is non standard, and you’ll have to grab my copy of prompt.py if you want this)</li> </ul>
<p>There’s a list of the information available for your prompt on the <a href="http://sjl.bitbucket.org/hg-prompt/">hg prompt project page</a>.</p> <p>If you want to see how to do it for git too, see <a href="http://en.newinstance.it/2010/05/23/git-autocompletion-and-enhanced-bash-prompt/">GIT autocompletion and enhanced bash prompt</a>.</p> <p>I have modifications for both my git prompt and hg prompt to make the status indicators more verbose (i.e. “[unstaged]” instead of “?”). If you want the changes, here’s my copy of <a href="https://github.com/jlfwong/dotfiles/blob/master/git-completion.bash">git-completion.bash</a> for git and <a href="https://github.com/jlfwong/dotfiles/blob/master/hg-prompt/prompt.py">prompt.py</a> for mercurial. My version of <code>prompt.py</code> also includes support for <a href="https://github.com/jlfwong/dotfiles/commit/16ff9758a03a1a06574b8909d9a9cd0ea8302b70">displaying the status of shelves</a> - I’ll send that upstream when I get the chance.</p> <h2 id="autocomplete">autocomplete</h2> <p>Autocompletion for mercurial is not a mercurial extension - it’s just a bash script. This will contextually complete things like flag names, file names, bookmarks and tags. For instance, if I do <code>hg up default/de[TAB]</code>, it will complete to <code>hg up default/default</code>.</p> <p>I can’t remember offhand where I originally got the autocomplete for mercurial, but you can grab it from my github repo: <a href="https://github.com/jlfwong/dotfiles/blob/master/hg-completion.bash">hg-completion.bash</a>. After you stick that file somewhere, just add a line like this to your <code>~/.bash_profile</code>:</p> <div><pre><code><span></span><span>source</span> ~/path/to/hg-completion.bash </code></pre></div> <h1 id="workflow">Workflow</h1> <p>This is my personal workflow at <a href="http://www.khanacademy.org">Khan Academy</a>. The workflows people use here vary wildly, as they jump between repositories, git and hg, python, javascript, server configuration, github, kiln, and phabricator.</p> <p>I’m using a mutable history workflow where each feature branch only ever contains one commit. You can read my explanation for why I do this in a previous post, <a href="http://jamie-wong.com/2012/05/25/an-argument-for-mutable-local-history/">An Argument for Mutable Local History</a>.</p> <p>There are two predominate review systems used at Khan Academy - <a href="http://phabricator.org/">Phabricator</a> and <a href="http://www.fogcreek.com/Kiln/">Kiln</a>. I was accustomed to using Phabricator from interning at Facebook, so that’s what I’ve been using and what I’ll be talking about here. The associated commandline tool is <a href="http://www.phabricator.com/docs/phabricator/article/Arcanist_User_Guide.html">arcanist</a>, so that’s what all of the <code>arc</code> commands below are about.</p> <h2 id="configuration">Configuration</h2> <h3 id="arcrc">.arcrc</h3> <p>By default, arcanist does not amend commits for mercurial. I wanted this ability, so I submitted a patch to support this. To make arcanist amend commits, add the following to your <code>~/.arcrc</code> (or to your project <code>.arcconfig</code>).</p> <div><pre><code><span></span><span>{</span> <span>"config"</span><span>:</span> <span>{</span> <span>"history.immutable"</span><span>:</span> <span>false</span> <span>}</span> <span>}</span> </code></pre></div> <p><strong>EDIT</strong>: The setting was originally <code>"immutable_history": false</code>, but it’s been changed to “history.immutable”.</p> <h3 id="project-hg-hgrc">project .hg/.hgrc</h3> <p>Many of my coworkers maintain their own complete remote clone of the main repository. I don’t do this, so my default remote is the repository shared by my team. In my project root, I have two remotes specified: one for my team and one for stable. Stable is what gets deployed to production.</p> <p>In the root of any mercurial repository, there should be a <code>.hg/.hgrc</code> file - mine looks something like this:</p> <pre><code>[paths] # Team remote default = https://example.com/path/to/khan/repositories/labs-cs # Production-deployable remote stable = https://example.com/path/to/khan/repositories/stable </code></pre> <h2 id="starting-a-feature">Starting a Feature</h2> <p>Before starting a feature, I always make sure I’m up to date, so I run the following:</p> <pre><code>hg pull hg up default/default </code></pre> <p>Note that <code>default/default</code> specifies the latest commit on the remote default.<br> The <code>default/default</code> tag is created by the <a href="http://jamie-wong.com/2012/07/11/my-mercurial-setup-and-workflow-at-khan-academy/#remotebranches">remotebranches</a> plugin. This is not always the same as just running <code>hg up</code> with no argument, since <code>hg pull</code> may not have pulled down any commits, and <code>tip</code> may refer to a commit that hasn’t been pushed.</p> <p>After filing or finding a Phabricator Maniphest task I want to work on, I create a new mercurial bookmark that includes the id of the task. For instance, I just put up a diff to fix an issue filed under T111 relating to an issue with timestamps. When I started working on the feature, I ran</p> <pre><code>hg book timestamp-T111 </code></pre> <p>By default, when you commit in mercurial, all bookmarks on the current commit will move forward. We’re trying to use bookmarks as lightweight local branches, so this isn’t the behaviour we want. To make it only move the currently active bookmark, add this to your <code>~/.hgrc</code>:</p> <pre><code>[bookmarks] track.current = True </code></pre> <p><strong>WARNING</strong>: Mercurial has both the concept of “branches” and “bookmarks”. Mercurial branches are <strong>not</strong> the same as git branches. When you commit on a mercurial branch, that branch is forever embedded into the commit, even when its pushed. I don’t have a good understanding of the branching mechanism in mercurial, but I would generally advise against using <code>hg branch</code> to work on feature branches. To look at all the options available for branching in mercurial, I suggest reading <a href="http://stevelosh.com/blog/2009/08/a-guide-to-branching-in-mercurial/">A Guide to Branching in Mercurial</a>.</p> <h2 id="saving-work-in-progress">Saving Work in Progress</h2> <p>Sometimes I’ll need to stop working on a feature to revise other features that have been reviewed or to review someone else’s changes. Since I’m using a mutable history workflow, I just commit the work on the bookmark to be amended later.</p> <pre><code>hg commit -m "WIP Fix timestamp issue" </code></pre> <p>When I need to return to this bookmark later, I can do so by running</p> <pre><code>hg up timestamp-T111 </code></pre> <p>Since I have <a href="http://jamie-wong.com/2012/07/11/my-mercurial-setup-and-workflow-at-khan-academy/#autocomplete">autocomplete</a> installed, I can just type <code>hg up times[TAB]</code>.</p> <h2 id="putting-a-diff-up-for-review">Putting a Diff up for Review</h2> <p>When I’m ready to put up a diff for review, I ensure that I have all my work committed in one commit. After I’ve tested that everything works, I run</p> <pre><code>arc diff </code></pre> <p>Then fill out the summary, test plan, associated tasks, who I want to review the change and anyone I want to cc on the diff in the editor that pops up from running <code>arc diff</code>.</p> <p>After I’m done, I save from the editor and exit, which completes the process and puts the diff up on Phabricator.</p> <p>If you did the <code>"immutable_history": false</code> configuration noted in the introduction of <a href="http://jamie-wong.com/2012/07/11/my-mercurial-setup-and-workflow-at-khan-academy/#Workflow">Workflow</a>, then the commit message will be amended to include all the information written during <code>arc diff</code>, along with the URL to the Phabricator Differential revision.</p> <h2 id="revising-the-diff">Revising the Diff</h2> <p>After someone has reviewed my changes and inevitably requested changes, I switch back to the bookmark, make the changes, amend the commit and send out an updated diff.</p> <pre><code>hg up timestamp-T111 # hack hack hack hg commit --amend arc diff </code></pre> <p>When I amend the commit, I don’t usually change the commit message. This means that messages like “fixed lint” or “split lines to fit in 80 characters” never enter the revision history. In my opinion, this is a good thing. For an explanation of why, I’ll refer again to my post <a href="http://jamie-wong.com/2012/05/25/an-argument-for-mutable-local-history/">An Argument for Mutable Local History</a>.</p> <p>When I run <code>arc diff</code> on a commit that’s already been put up for review, it will bring up an editor asking what changed. This tells reviewers that you’ve “fixed lint” without sticking it in the revision history forever.</p> <h2 id="pushing-the-commit">Pushing the Commit</h2> <p>After the diff has been approved by one or more of the reviewers, it’s time to share that change with the rest of my team. At Khan Academy, each team has their own remote clone of the repository (this is what happens when you hit “branch” in Kiln).</p> <p>To land the change, I make sure I’m up to date with the remote, and rebase my change onto default. Rebasing requires that the builtin <a href="http://jamie-wong.com/2012/07/11/my-mercurial-setup-and-workflow-at-khan-academy/#rebase">rebase</a> extension be enabled. I rebase instead of merge in order to maintain a linear commit history, which is much easier to read and work with.</p> <pre><code>hg pull hg up timestamp-T111 hg rebase -d default/default </code></pre> <p>At this point, it’s possible I’ve run into some merge conflicts. I can get a list of these conflicts using</p> <pre><code>hg resolve -l </code></pre> <p>Unresolved conflicts will have “U” before the filename. “R” means there was a conflict, but it was automatically resolved.</p> <p>Once I’ve gone through all the unresolved conflicts, I can mark them as resolved using</p> <pre><code>hg resolve --mark [FILENAMES] </code></pre> <p>From here, I can continue the rebase (<em>without</em> committing) by running</p> <pre><code>hg rebase --continue </code></pre> <p>Once my commit has been successfully rebased, I run through the test plan described when I first ran <code>arc diff</code> to ensure that everything is still working. I also run the entire suite of automated tests and ensure everything passes lint. For Khan Academy, that’s done by running <code>make check</code>.</p> <p>Now I double check which commit is going out by running</p> <pre><code>hg outgoing -r . </code></pre> <p>Now I’m ready to push, which I do by running</p> <pre><code>hg push -r . </code></pre> <p>Note that the <code>-r .</code> is required here. <code>hg push</code> with no arguments will attempt to push all commits not in the remote (i.e. will push all commits listed by <code>hg outgoing</code> with no arguments). This might attempt to push many different features, some of which have not yet been reviewed. This will fail with a message from mercurial about creating multiple heads on the remote. If you specify a revision with <code>-r</code>, it will only push revisions that are ancestors of the commit specified.</p> <p>Since I always want to use <code>-r .</code> whenever I push, I have <code>hg nudge</code> aliased to <code>hg push</code>. Thanks to <a href="http://david-hu.com/">David Hu</a> for the alias.</p> <div><pre><code><span></span><span>[alias]</span> <span>nudge</span> <span>=</span> <span>push --rev .</span> </code></pre></div> <p>If everything went okay, I close the Phabricator Maniphest task and delete the bookmark, since it’s no longer needed.</p> <pre><code>arc close T111 hg book -d timestamp-T111 </code></pre> <h2 id="merging-changes-into-stable">Merging Changes into Stable</h2> <p>When my team decides that we have a group of features ready to go into production, we merge our branch with stable and push to both remotes.</p> <p>This one has to be a merge — as opposed to a rebase — because any changes in either stable or the team remote have already been shared with other people.<br><strong>Never rebase shared commits</strong>. Mercurial will actually attempt to stop you from rebasing any commit in the “public” phase.</p> <p>The merge process will look something like this:</p> <pre><code># Sync with remote hg pull hg pull stable # Merge team branch with stable hg up stable/default hg merge default/default # Deal with any conflicts using hg resolve hg commit -m 'Merge stable + labs-cs' # Test to make sure everything is working # Push changes hg push -r . # Push to team remote hg push stable -r . # Push to stable remote </code></pre> <h2 id="reviewing-others-changes">Reviewing Others’ Changes</h2> <p>When someone lists me as a reviewer, I need to pull down their changes to test them out locally. First, I revert my working copy to be the same as the remote.</p> <pre><code>hg pull hg up default/default </code></pre> <p>Usually the person requesting review is on my team. If they weren’t, I’d instead do</p> <pre><code>hg pull stable hg up stable/default </code></pre> <p>Then I apply the patch I’m reviewing. With arcanist, this is done by doing something like</p> <pre><code>arc patch D381 </code></pre> <p>I play around with it and test it out, then when I’m done reviewing, I clean my local directory state. Assuming you have the <a href="http://jamie-wong.com/2012/07/11/my-mercurial-setup-and-workflow-at-khan-academy/#purge">purge</a> extension, this is done with</p> <pre><code>hg clean hg up -C . </code></pre> <p>or if you crated the nuke alias described in the <a href="http://jamie-wong.com/2012/07/11/my-mercurial-setup-and-workflow-at-khan-academy/#purge">purge</a> section above, just</p> <pre><code>hg nuke </code></pre> <h1 id="things-i-m-missing">Things I’m Missing</h1> <p>Overall, I’m pretty happy with my mercurial setup. It provides rough analogs of everything I’m used to using in git, and that’s all I really need.</p> <p>There are, however, a few things I had while working with git that I’m still lacking in mercurial. Some of these may be due to ignorance, some of them due to actual non-existence.</p> <h2 id="git-grep">Git Grep</h2> <p><code>hg grep</code> and <code>git grep</code> do very different things. From the man pages, <code>git grep</code> “Look[s] for specified patterns in the tracked files in the work tree”, whereas <code>hg grep</code> “Search[es] revisions of files for a regular expression. […] It searches repository history, not the working directory.”</p> <p>For now I’ve been using <a href="http://betterthangrep.com/">ack</a>, which works well enough. The real advantage of <code>git grep</code> is that it ignores all files ignored by the <code>.gitignore</code> in the current repository, and the files you want ignored might vary project by project. It’s also really fast.</p> <h2 id="fugitive-vim-plugin">Fugitive - Vim Plugin</h2> <p>While working in git repositories, I have <a href="https://github.com/tpope/vim-fugitive/">fugitive</a> - an awesome vim plugin written by <a href="https://twitter.com/tpope">Tim Pope</a>. It provides a lot of really cool ways of interacting with the repository. Two of my absolute favorites are <code>:GDiff</code> and <code>:GBlame</code>.</p> <p>I won’t talk about it much here, but if you’re using vim and work in git repositories, you should seriously check it out.</p> <p><strong>EDIT</strong>: I <em>sort</em> of made it: <a href="https://github.com/jlfwong/vim-mercenary">vim-mercenary</a>.</p> </article></div><br>
<br>
<br>
<br>
<br>
<br>
via Zero Wind :: Jamie Wong http://jamie-wong.com<br>
<br>
June 22, 2019 at 11:15PM
